File: program/program/implement-jump-search/implement_jump_search.dart
import 'dart:math';

int jumpSearch(List<int> array, int x) {
  int n = array.length;
  int blockSize = sqrt(n).toInt();
  int prev = 0;

  // Jump to the right block
  while (array[min(blockSize, n) - 1] < x) {
    prev = blockSize;
    blockSize += sqrt(n).toInt();
    if (prev >= n) {
      return -1; // Element not found
    }
  }

  // Linear search within the block
  while (array[prev] < x) {
    prev++;
    if (prev == min(blockSize, n)) {
      return -1; // Element not found
    }
  }

  if (array[prev] == x) {
    return prev; // Element found at index prev
  }

  return -1; // Element not found
}

void main() {
  List<int> list = [1, 2, 3, 4, 5];
  int value = 4;

  int result = jumpSearch(list, value);

  if (result != -1) {
    print('Element $value found at index $result');
  } else {
    print('Element $value not found in the list');
  }
}

File: program/program/implement-jump-search/implement_jump_search.go
package main

import (
	"fmt"
	"math"
)

func jumpSearch(arr []int, n int, x int) int {
	blockSize := int(math.Sqrt(float64(n)))
	prev := 0

	for arr[min(blockSize-1, n-1)] < x {
		prev = blockSize
		blockSize += int(math.Sqrt(float64(n)))
		if prev >= n {
			return -1
		}
	}

	for arr[prev] < x && prev < min(blockSize, n) {
		prev++
	}

	if arr[prev] == x {
		return prev
	}

	return -1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	arr := []int{1, 2, 3, 4, 5}
	x := 5

	index := jumpSearch(arr, len(arr), x)

	if index != -1 {
		fmt.Println("Value found at index:", index)
	} else {
		fmt.Println("Value not found in the list")
	}
}
